File: 437.Path-Sum-III.go
// https://leetcode.com/problems/path-sum-iii/
//
// algorithms
// Easy (41.38%)
// Total Accepted:    86,391
// Total Submissions: 208,789
// beats 83.33% of golang submissions

package leetcode

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var pathSumRes int

func pathSum(root *TreeNode, sum int) int {
	if root == nil {
		return 0
	}

	pathSumRes = 0

	resursive(root, sum)

	return pathSumRes
}

func resursive(node *TreeNode, sum int) []int {
	if node.Left == nil && node.Right == nil {
		if node.Val == sum {
			pathSumRes++
		}
		return []int{node.Val}
	}

	res := []int{node.Val}
	if node.Val == sum {
		pathSumRes++
	}
	if node.Left != nil {
		leftArr := resursive(node.Left, sum)
		for idx, n := range leftArr {
			leftArr[idx] = n + node.Val
			if leftArr[idx] == sum {
				pathSumRes++
			}
		}
		res = append(res, leftArr...)
	}
	if node.Right != nil {
		rightArr := resursive(node.Right, sum)
		for idx, n := range rightArr {
			rightArr[idx] = n + node.Val
			if rightArr[idx] == sum {
				pathSumRes++
			}
		}
		res = append(res, rightArr...)
	}

	return res
}

File: 437.Path-Sum-III.py
# https://leetcode.com/problems/path-sum-iii/
#
# algorithms
# Easy (43.45%)
# Total Accepted:    117,973
# Total Submissions: 271,533


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        res = [0]

        def recursive(node, pre_sum_list):
            if not node:
                return

            for idx, n in enumerate(pre_sum_list):
                tmp = n + node.val
                pre_sum_list[idx] = tmp
                if tmp == sum:
                    res[0] += 1

            if node.val == sum:
                res[0] += 1

            recursive(node.left, pre_sum_list + [node.val])
            recursive(node.right, pre_sum_list + [node.val])

        recursive(root, [])

        return res[0]
